File: client/src/models/environments.ts
import { immerable } from "immer";
import { serializable } from "@/utils/serializable";
import { Variable } from "@/models/variables";
import { Connection } from "@/models/connections";

export class Environment {
  [immerable] = true;

  @serializable("string", { required: true, format: "uuid" })
  id!: string;

  @serializable("string", { required: true, format: "identifier" })
  name!: string;

  @serializable("string")
  description?: string;

  @serializable("string")
  ref?: string;

  @serializable("object", { factory: Connection })
  connections?: Connection[];

  @serializable("object", { factory: Variable })
  variables?: Variable[];

  constructor(object?: Partial<Environment>) {
    Object.assign(this, object);
    this.connections = object?.connections?.map((c) => new Connection(c));
    this.variables = object?.variables?.map((v) => new Variable(v));
  }
}

File: agent/src/models/environments.rs
use crate::models::variables::Variable;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct Environment {
    pub id: Uuid,

    pub parent_id: Uuid,

    pub owner_user_id: Uuid,

    pub name: String,

    #[serde(skip_serializing_if = "String::is_empty")]
    pub description: String,

    #[serde(rename = "ref")]
    pub connections: Vec<Uuid>,
    pub variables: Vec<Variable>,
}
